File: script-frontend/src/image-processing/ocr_extractor.py
#!/usr/bin/env python3
"""
ocr_extractor.py

Persistent OCR worker process using EasyOCR (cost-effective alternative to Google Cloud Vision).

Protocol (newline-delimited JSON over stdin/stdout):
  Request:  {"images": ["/path/to/img1.jpg", "/path/to/img2.jpg"]}
  Response: [{"image_path": "...", "text": "...", "words": [...], "confidence": 0.95}, ...]

  Special:  {"cmd": "ping"}   ->  {"status": "ok"}
            {"cmd": "quit"}   ->  process exits

The model is loaded once at startup and kept in memory for all subsequent requests.

Legacy mode (CLI arguments) is still supported for backward compatibility:
  python ocr_extractor.py <image_path> [image_path2 ...]
"""
import sys
import json
import ssl
import os
import signal
import faulthandler
from pathlib import Path
from io import StringIO

import cv2
import numpy as np

# ── Suppress all warnings and noisy library output ──────────────────────────
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
import warnings
warnings.filterwarnings('ignore')

# Reduce PyTorch memory footprint: disable gradients and limit internal threads
import torch
torch.set_grad_enabled(False)
torch.set_num_threads(1)

# Handle SSL certificate issues on macOS
try:
    _create_unverified_https_context = ssl._create_unverified_context
except AttributeError:
    pass
else:
    ssl._create_default_https_context = _create_unverified_https_context

# ── Import EasyOCR (suppressing its chatty output) ──────────────────────────
_real_stdout = sys.stdout
_real_stderr = sys.stderr

# Dump a Python-level stack trace to stderr on fatal signals (SIGSEGV/SIGBUS/SIGABRT)
faulthandler.enable()

def _emergency_exit(signum, frame):
    """Flush stdout so the Node.js readline closes cleanly, then exit immediately."""
    try:
        _real_stdout.flush()
    except Exception:
        pass
    os._exit(1)  # skip normal Python cleanup which can deadlock in multi-threaded PyTorch

signal.signal(signal.SIGBUS, _emergency_exit)
signal.signal(signal.SIGTERM, _emergency_exit)

try:
    sys.stdout = StringIO()
    sys.stderr = StringIO()
    import easyocr
    sys.stdout = _real_stdout
    sys.stderr = _real_stderr
except ImportError:
    sys.stdout = _real_stdout
    sys.stderr = _real_stderr
    print("Error: easyocr module not found. Install it with: pip install easyocr", file=sys.stderr)
    os._exit(1)
except Exception:
    sys.stdout = _real_stdout
    sys.stderr = _real_stderr
    os._exit(1)

# ── Model management ────────────────────────────────────────────────────────
_reader = None

def get_reader():
    global _reader
    if _reader is None:
        buf_out, buf_err = StringIO(), StringIO()
        sys.stdout, sys.stderr = buf_out, buf_err
        try:
            _reader = easyocr.Reader(['en'], gpu=False, verbose=False)
        finally:
            sys.stdout = _real_stdout
            sys.stderr = _real_stderr
            buf_out.close()
            buf_err.close()
    return _reader

# ── Core OCR function ───────────────────────────────────────────────────────
def extract_text(image_path):
    """Extract all text from an image using EasyOCR."""
    buf_out, buf_err = StringIO(), StringIO()
    sys.stdout, sys.stderr = buf_out, buf_err
    try:
        ocr_reader = get_reader()
        with torch.inference_mode():
            results = ocr_reader.readtext(str(image_path))
        full_text = ' '.join([d[1] for d in results])
        return {
            'image_path': str(image_path),
            'text': full_text,
            'words': [d[1] for d in results],
            'confidence': sum(d[2] for d in results) / len(results) if results else 0,
        }
    except Exception as e:
        return {'image_path': str(image_path), 'text': '', 'error': str(e)}
    finally:
        sys.stdout = _real_stdout
        sys.stderr = _real_stderr
        buf_out.close()
        buf_err.close()


def process_images(image_paths):
    """Process a list of image paths and return results."""
    results = []
    for img_path in image_paths:
        if not Path(img_path).exists():
            results.append({'image_path': img_path, 'text': '', 'error': 'File not found'})
            continue
        results.append(extract_text(img_path))
    return results

# ── Orientation detection ──────────────────────────────────────────────────
# Ported from orient_cards.py — reuses the already-loaded EasyOCR model
# instead of spawning a separate process that loads a duplicate (~300MB).

MAX_SIDE_FOR_OCR = 1024
EARLY_EXIT_AVG_CONF = 0.90
EARLY_EXIT_MIN_DETECTIONS = 3
MIN_MEANINGFUL_CONF = 0.3

ROTATIONS = {
    0:   None,
    90:  cv2.ROTATE_90_CLOCKWISE,
    180: cv2.ROTATE_180,
    270: cv2.ROTATE_90_COUNTERCLOCKWISE,
}


def downscale(img, max_side=MAX_SIDE_FOR_OCR):
    h, w = img.shape[:2]
    if max(h, w) <= max_side:
        return img
    scale = max_side / max(h, w)
    return cv2.resize(img, (int(w * scale), int(h * scale)), interpolation=cv2.INTER_AREA)


def ocr_score(img, reader):
    """Run EasyOCR on img and return (sum_confidence, avg_confidence, num_detections)."""
    buf_out, buf_err = StringIO(), StringIO()
    sys.stdout, sys.stderr = buf_out, buf_err
    try:
        with torch.inference_mode():
            results = reader.readtext(img)
    finally:
        sys.stdout = _real_stdout
        sys.stderr = _real_stderr
        buf_out.close()
        buf_err.close()
    meaningful = [d for d in results if d[2] >= MIN_MEANINGFUL_CONF]
    if not meaningful:
        return 0.0, 0.0, 0
    total_conf = sum(d[2] for d in meaningful)
    avg_conf = total_conf / len(meaningful)
    return total_conf, avg_conf, len(meaningful)


def detect_orientation(image_path, reader):
    """Detect correct orientation, apply rotation in-place if needed."""
    img = cv2.imread(str(image_path))
    if img is None:
        return {
            'image_path': str(image_path),
            'rotation_applied': 0,
            'confidence': 0.0,
            'text_detection_count': 0,
            'scores': {},
            'error': 'Could not read image',
        }

    small = downscale(img)
    scores = {}

    total_0, avg_0, n_0 = ocr_score(small, reader)
    scores['0'] = round(total_0, 3)

    if avg_0 >= EARLY_EXIT_AVG_CONF and n_0 >= EARLY_EXIT_MIN_DETECTIONS:
        return {
            'image_path': str(image_path),
            'rotation_applied': 0,
            'confidence': round(avg_0, 3),
            'text_detection_count': n_0,
            'scores': scores,
        }

    best_angle = 0
    best_total = total_0
    best_detections = n_0

    for angle in [90, 180, 270]:
        rotated = cv2.rotate(small, ROTATIONS[angle])
        total, avg, n = ocr_score(rotated, reader)
        scores[str(angle)] = round(total, 3)
        if total > best_total:
            best_total = total
            best_angle = angle
            best_detections = n

    if best_total < MIN_MEANINGFUL_CONF:
        return {
            'image_path': str(image_path),
            'rotation_applied': 0,
            'confidence': 0.0,
            'text_detection_count': 0,
            'scores': scores,
        }

    if best_angle != 0:
        rotated_full = cv2.rotate(img, ROTATIONS[best_angle])
        cv2.imwrite(str(image_path), rotated_full, [cv2.IMWRITE_JPEG_QUALITY, 95])

    return {
        'image_path': str(image_path),
        'rotation_applied': best_angle,
        'confidence': round(best_total, 3),
        'text_detection_count': best_detections,
        'scores': scores,
    }


def process_orient(image_paths):
    """Orient a batch of images and return results."""
    reader = get_reader()
    results = []
    for img_path in image_paths:
        if not Path(img_path).exists():
            results.append({
                'image_path': str(img_path),
                'rotation_applied': 0,
                'confidence': 0.0,
                'text_detection_count': 0,
                'scores': {},
                'error': 'File not found',
            })
            continue
        results.append(detect_orientation(img_path, reader))
    return results


# ── Persistent worker loop ──────────────────────────────────────────────────
def run_worker():
    """Read JSON requests from stdin, write JSON responses to stdout (one per line).

    This keeps the process (and the loaded model) alive between calls,
    avoiding the ~200-400 MB allocation spike on every invocation.
    """
    # Eagerly load the model so the first request is fast
    get_reader()

    # Signal readiness
    _real_stdout.write(json.dumps({"status": "ready"}) + "\n")
    _real_stdout.flush()

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            request = json.loads(line)
        except json.JSONDecodeError:
            _real_stdout.write(json.dumps({"error": "Invalid JSON"}) + "\n")
            _real_stdout.flush()
            continue

        # Handle special commands
        cmd = request.get("cmd")
        if cmd == "ping":
            _real_stdout.write(json.dumps({"status": "ok"}) + "\n")
            _real_stdout.flush()
            continue
        if cmd == "quit":
            break
        if cmd == "orient":
            image_paths = request.get("images", [])
            if not image_paths:
                _real_stdout.write(json.dumps([]) + "\n")
                _real_stdout.flush()
                continue
            results = process_orient(image_paths)
            _real_stdout.write(json.dumps(results) + "\n")
            _real_stdout.flush()
            continue

        # Process images (default: OCR text extraction)
        image_paths = request.get("images", [])
        if not image_paths:
            _real_stdout.write(json.dumps([]) + "\n")
            _real_stdout.flush()
            continue

        results = process_images(image_paths)
        _real_stdout.write(json.dumps(results) + "\n")
        _real_stdout.flush()

# ── Entry point ─────────────────────────────────────────────────────────────
def main():
    # If CLI arguments are provided, run in legacy one-shot mode
    if len(sys.argv) > 1:
        try:
            results = process_images(sys.argv[1:])
            print(json.dumps(results))
            sys.stdout.flush()
            os._exit(0)
        except Exception:
            print(json.dumps([{'error': 'Failed to process images'}]))
            os._exit(1)

    # No arguments → persistent worker mode
    run_worker()
    os._exit(0)  # skip Py_FinalizeEx — avoids PyTorch/C-extension segfault during module cleanup

if __name__ == "__main__":
    main()

File: script-frontend/src/image-processing/ocr-extractor.ts
import { spawn, type ChildProcessWithoutNullStreams } from 'child_process';
import path from 'path';
import { fileURLToPath } from 'url';
import { createInterface, type Interface as ReadlineInterface } from 'readline';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

type OCRExtractResult = {
  image_path: string;
  text: string;
  words?: string[];
  confidence?: number;
  error?: string;
};

export type OrientResult = {
  image_path: string;
  rotation_applied: number;
  confidence: number;
  text_detection_count: number;
  scores: Record<string, number>;
  error?: string;
};

// ── Configuration ───────────────────────────────────────────────────────────

const MAX_RETRIES = 2;
const RESTART_DELAY_MS = 1_000;
const REQUEST_TIMEOUT_MS = 120_000; // 2 min per request (model init on first call can be slow)
const WORKER_STARTUP_TIMEOUT_MS = 180_000; // 3 min max for model load

// ── Persistent OCR worker management ────────────────────────────────────────

let worker: ChildProcessWithoutNullStreams | null = null;
let workerReady = false;
let readline: ReadlineInterface | null = null;
let spawnPromise: Promise<void> | null = null;

/** Queue to serialize requests to the worker (only one in-flight at a time). */
type QueuedRequest = {
  payload: object;
  resolve: (line: string) => void;
  reject: (err: Error) => void;
};
const requestQueue: QueuedRequest[] = [];
let activeRequest: QueuedRequest | null = null;

function getEnv(): NodeJS.ProcessEnv {
  const threads = process.env.OCR_THREADS || '2';
  return {
    ...process.env,
    // Limit BLAS / OpenMP threads to reduce memory footprint
    OMP_NUM_THREADS: threads,
    MKL_NUM_THREADS: threads,
    OPENBLAS_NUM_THREADS: threads,
    NUMEXPR_NUM_THREADS: threads,
    // Prevent PyTorch from caching freed allocations (reduces peak memory)
    PYTORCH_NO_CUDA_MEMORY_CACHING: '1',
    // Disable macOS nano-zone allocator – known to interact badly with the
    // Tahoe memory-compressor bug that causes SIGBUS crashes.
    MALLOC_NANO_ZONE: '0',
    // macOS Tahoe: Objective-C runtime crashes in subprocess when ObjC frameworks
    // (used internally by PyTorch/MPS) are initialized across fork boundaries.
    OBJC_DISABLE_INITIALIZE_FORK_SAFETY: 'YES',
  };
}

/**
 * Spawn (or re-spawn) the persistent Python OCR worker.
 * Resolves once the worker prints its {"status":"ready"} line.
 */
function spawnWorker(): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    const scriptPath = path.join(__dirname, 'ocr_extractor.py');
    const venvPython = path.join(__dirname, '..', '..', 'venv', 'bin', 'python3');

    // Kill any lingering process
    killWorker();

    const child = spawn(venvPython, [scriptPath], {
      stdio: ['pipe', 'pipe', 'pipe'],
      env: getEnv(),
    });

    worker = child;
    workerReady = false;

    // Accumulate stderr so we can log it if the worker crashes unexpectedly
    let stderrBuffer = '';
    child.stderr.on('data', (chunk: Buffer) => {
      stderrBuffer += chunk.toString();
      // Keep last 4 KB to avoid unbounded growth
      if (stderrBuffer.length > 4096) stderrBuffer = stderrBuffer.slice(-4096);
    });

    const rl = createInterface({ input: child.stdout });
    readline = rl;

    // First line must be the {"status":"ready"} handshake
    let gotReady = false;

    // Reject if the worker never emits ready (e.g. crashes during model load)
    const startupTimer = setTimeout(() => {
      if (!gotReady) {
        killWorker();
        reject(new Error('OCR worker startup timed out'));
      }
    }, WORKER_STARTUP_TIMEOUT_MS);

    rl.on('line', (line: string) => {
      if (!gotReady) {
        // Consume the ready signal
        try {
          const msg = JSON.parse(line);
          if (msg?.status === 'ready') {
            gotReady = true;
            workerReady = true;
            clearTimeout(startupTimer);
            resolve();
            return;
          }
        } catch {
          // ignore non-JSON during startup
        }
        return;
      }

      // Subsequent lines are responses to requests
      if (activeRequest) {
        const req = activeRequest;
        activeRequest = null;
        req.resolve(line);
        // Process next queued request
        processNextRequest();
      }
    });

    child.on('error', (err) => {
      workerReady = false;
      if (!gotReady) {
        clearTimeout(startupTimer);
        reject(new Error(`Failed to start OCR worker: ${err.message}`));
      }
      rejectPending(new Error(`OCR worker error: ${err.message}`));
    });

    child.on('close', (code, signal) => {
      workerReady = false;
      worker = null;
      readline = null;
      if (code !== 0 || signal) {
        console.error(`[OCR worker] exited code=${code} signal=${signal}`);
        if (stderrBuffer.trim()) {
          console.error('[OCR worker stderr]:\n' + stderrBuffer.trim());
        }
      }
      if (!gotReady) {
        clearTimeout(startupTimer);
        reject(new Error(`OCR worker exited during startup (code=${code}, signal=${signal})`));
      }
      rejectPending(new Error(`OCR worker crashed (code=${code}, signal=${signal})`));
    });
  });
}

function killWorker() {
  if (worker) {
    try {
      worker.stdin.end();
      worker.kill();
    } catch {
      // already dead
    }
    worker = null;
    workerReady = false;
    readline = null;
  }
}

function rejectPending(err: Error) {
  // Reject the active request
  if (activeRequest) {
    const req = activeRequest;
    activeRequest = null;
    req.reject(err);
  }
  // Reject all queued requests
  while (requestQueue.length > 0) {
    const req = requestQueue.shift()!;
    req.reject(err);
  }
}

/**
 * Ensure the worker is alive and ready. Spawns it on first call or after a crash.
 */
async function ensureWorker(): Promise<void> {
  if (worker && workerReady) return;
  if (!spawnPromise) {
    spawnPromise = spawnWorker().finally(() => {
      spawnPromise = null;
    });
  }
  await spawnPromise;
}

/**
 * Flush the next queued request to the worker's stdin.
 * Only called when no request is currently in-flight.
 */
function processNextRequest(): void {
  if (activeRequest || requestQueue.length === 0) return;
  if (!worker || !workerReady) {
    // Reject everything – caller will restart the worker and retry
    while (requestQueue.length > 0) {
      const req = requestQueue.shift()!;
      req.reject(new Error('OCR worker not ready'));
    }
    return;
  }

  const req = requestQueue.shift()!;
  activeRequest = req;

  const line = JSON.stringify(req.payload) + '\n';
  worker.stdin.write(line, (err) => {
    if (err) {
      activeRequest = null;
      req.reject(new Error(`Failed to write to OCR worker: ${err.message}`));
      // Try the next request
      processNextRequest();
    }
  });
}

/**
 * Send a single request to the worker and wait for the response line.
 * Requests are serialized so only one is in-flight at a time.
 */
function sendRequest(payload: object): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    if (!worker || !workerReady) {
      reject(new Error('OCR worker not ready'));
      return;
    }

    requestQueue.push({ payload, resolve, reject });

    // If nothing is in-flight, kick off processing immediately
    if (!activeRequest) {
      processNextRequest();
    }
  });
}

// ── Public API ──────────────────────────────────────────────────────────────

/**
 * Extract text from images using the persistent EasyOCR worker.
 * Automatically (re)starts the worker and retries on transient crashes.
 * @param imagePaths - Array of image paths to extract text from
 * @returns Array of extracted text results
 */
export async function extractTextWithOCR(imagePaths: string[]): Promise<OCRExtractResult[]> {
  let lastError: Error | undefined;

  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    try {
      await ensureWorker();

      // Race the request against a timeout
      const responseLine = await Promise.race([
        sendRequest({ images: imagePaths }),
        new Promise<never>((_, reject) =>
          setTimeout(() => reject(new Error('OCR request timed out')), REQUEST_TIMEOUT_MS),
        ),
      ]);

      return JSON.parse(responseLine) as OCRExtractResult[];
    } catch (err) {
      lastError = err as Error;

      // If the worker crashed, kill it so ensureWorker will restart on next attempt
      killWorker();

      // Don't retry if we can't even start the process
      if (lastError.message.startsWith('Failed to start OCR worker')) {
        throw lastError;
      }

      // Brief pause before retry to let memory pressure subside
      if (attempt < MAX_RETRIES) {
        await new Promise((r) => setTimeout(r, RESTART_DELAY_MS * (attempt + 1)));
      }
    }
  }

  throw lastError!;
}

/**
 * Extract text from a single image
 * @param imagePath - Path to the image
 * @returns Extracted text and metadata
 */
export async function extractTextFromImage(imagePath: string): Promise<OCRExtractResult> {
  const results = await extractTextWithOCR([imagePath]);
  return results[0];
}

/**
 * Get combined text from multiple images
 * @param imagePaths - Array of image paths
 * @returns Combined text from all images
 */
export async function extractTextFromImages(imagePaths: string[]): Promise<string> {
  const results = await extractTextWithOCR(imagePaths);
  return results.map((r) => r.text).join(' ');
}

/**
 * Detect and correct orientation of cropped card images using the persistent
 * EasyOCR worker. Images are rotated in-place if needed.
 *
 * This replaces the standalone orient_cards.py script, avoiding a duplicate
 * ~300MB EasyOCR model load.
 */
export async function orientImages(imagePaths: string[]): Promise<OrientResult[]> {
  if (imagePaths.length === 0) return [];

  let lastError: Error | undefined;

  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    try {
      await ensureWorker();

      const responseLine = await Promise.race([
        sendRequest({ cmd: 'orient', images: imagePaths }),
        new Promise<never>((_, reject) =>
          setTimeout(() => reject(new Error('Orient request timed out')), REQUEST_TIMEOUT_MS),
        ),
      ]);

      return JSON.parse(responseLine) as OrientResult[];
    } catch (err) {
      lastError = err as Error;
      killWorker();

      if (lastError.message.startsWith('Failed to start OCR worker')) {
        throw lastError;
      }

      if (attempt < MAX_RETRIES) {
        await new Promise((r) => setTimeout(r, RESTART_DELAY_MS * (attempt + 1)));
      }
    }
  }

  throw lastError!;
}

/**
 * Gracefully shut down the OCR worker. Call this when your app is exiting.
 */
export function shutdownOCR(): void {
  if (worker && workerReady) {
    try {
      worker.stdin.write(JSON.stringify({ cmd: 'quit' }) + '\n');
    } catch {
      // ignore
    }
  }
  killWorker();
}

// Clean up on process exit
process.on('exit', () => killWorker());
process.on('SIGINT', () => {
  killWorker();
  process.exit(130);
});
process.on('SIGTERM', () => {
  killWorker();
  process.exit(143);
});